Multiarch builds

fgiorgetti opened the pull request.

grs commented on Dockerfile.ci-test:
```diff
@@ -1,4 +1,6 @@
-FROM golang:1.19 AS builder
+FROM --platform=$BUILDPLATFORM golang:1.19 AS builder
+
+ARG TARGETOS TARGETARCH
```
I get a build error from these lines using podman, I think it is expecting a separate line per ARG definition.

fgiorgetti replied:
Updated.
I was able to manually build and publish multiarch images with podman, but the commands
differ a bit from docker.

nluaces commented:
Running it locally with docker 23.0.2,  I'm getting this error:
```
» make docker-build                                                                                                                                                      nluaces@thinkpad
docker buildx build --no-cache --platform linux/amd64,linux/s390x -t quay.io/nluaces/skupper-tests -f Dockerfile.ci-test .
[+] Building 0.0s (0/0)                                                                                                                                                                                                                 
ERROR: multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")
make: *** [Makefile:46: docker-build-test-image] Error 1
```
Just for the test, I'm removing one of the platforms and everything works. Which local environment do you have to test the two platforms?

fgiorgetti commented:
> Running it locally with docker 23.0.2, I'm getting this error:
> 
> ```
> » make docker-build          
> docker buildx build --no-cache --platform linux/amd64,linux/s390x -t quay.io/nluaces/skupper-tests -f Dockerfile.ci-test .
> [+] Building 0.0s (0/0)                                                                                                                                                                                                                 
> ERROR: multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")
> make: *** [Makefile:46: docker-build-test-image] Error 1
> ```
> 
> Just for the test, I'm removing one of the platforms and everything works. Which local environment do you have to test the two platforms?

Try creating the builder instance using: `docker buildx create --use --name skupper-buildx` (like done in the CI).
Then see if it works for you (my docker version is a bit older than yours and seems to be fine). 

To test it, I have a local s390x VM  where I am able to run the CLI and validate podman sites (using the s390x images).

To run it, I am using QEMU and to be able to create a VM for this architecture (on my Fedora), I had to install qemu-system-s390x.

sonbui00 commented:
@fgiorgetti can you resolve this? let me know if you need a help on this

fgiorgetti commented:
> @fgiorgetti can you resolve this? let me know if you need a help on this

Hello @sonbui00. Before we can actually rebase and merge this PR, we need to get
the [skupper-router](https://github.com/skupperproject/skupper-router/) ready for multiarch as well.

There is an [ongoing PR](https://github.com/skupperproject/skupper-router/pull/1100) for it as well.
We expect to make some progress shortly.

fgiorgetti commented:
@grs @nluaces I have just rebased and adjusted the image build process to include the controller-podman image.
Could you please give it a one more try?

fgiorgetti commented:
@grs @nluaces @ajssmith @c-kruse I have removed s390x as a target platform for images,
as the router image is not yet available for it.
